### ISSUE TITLE
Resetting stats in NeighborSearch.

### ISSUE THREAD

**Issue by MarcosPividori** [mlpack/mlpack#672]
In NeighborSearch, when the method Search(Tree* queryTree, ...) uses the given tree, it is assuming that stats there were intialized to their initial values.
Shouldn't we add a comment about this to the documentation?
I mean, if we do: 

  KNN knn; 
  KNN::Tree queryTree(querySet);
  knn.Train(someReferenceSet); 
  knn.Search(queryTree,.....); 
  knn.Train(someDifferentReferenceSet);

The last search will probably give incorrect results, because the queryTree holds old unrelated statistic values for the bounds. I mean, it uses bound values from the first referenceSet to search in the second referenceSet.

**Comment by MarcosPividori** [comment]
@sumedhghaisas @rcurtin 

After a code review, I think we have this 2 options:

- Only add a comment to the docummentation of the method Search(Tree* queryTree, ...), mentioning that the queryTree's stats should be resetted before calling Search again with the same queryTree and a different reference dataset.

- Modify the method Search(Tree* queryTree, ...), to reset all stats of the tree at the beginning. This could slow down the calculation a bit. For example, in NSModel, the class creates a new tree and then call Search. In that case, resetting the stats wouldn't be necessary, because the tree is new...

**Comment by rcurtin** [comment]
I think the issue here is not lack of documentation but instead that the `treeNeedsReset` member is used to check if a search has already been done, but it resets the statistics in the reference tree, not the query tree. If we change this code to work with the query tree instead of the reference tree, I think there is no issue.

**Comment by MarcosPividori** [comment]
The problem is that the query tree is provided as an argument every time Search(Tree* queryTree, ...) method is called, so we can not track when it is necessary to reset the stats there. We don't know where that query tree comes from... if it is a new one, or was previously used for something different...

This is what I understand, let me know if you agree:

- The problem appears when we use the same query tree with different reference sets. I mean, when reusing outdated stats in a different scenario.

- When we do monochromatic search, the reference set and query set are always the same, so we never reach the erroneous situation. So, treeNeedsReset, as used now, is irrelevant. I mean, NeighborSearch class reset the stats in the reference tree every time we repeat a monochromatic search. It is not necessary, because the reference set never changes. If it changes, a new tree is built, with new stats, so there isn't any problem.

- On the other hand, we should reset stats when using the same query tree with different reference datasets. But we don't know that information, because the query tree is not part of the state of NeighborSearch class. It is provided as an argument every time Search(Tree* queryTree, ...) method is called, so we can not track when it is necessary to reset the stats there. We don't know where that query tree comes from... if it is a new one, or was previously used for something different...

Because of that, I think we should take one of the approaches:
+ allways reset the provided query tree. This way we are sure everything is ok, but slows down the calculation a bit.
+ never reset, and add a comment mentioned this, so the user can decide if it is necessary.

**Comment by MarcosPividori** [comment]
Sorry, I confused about `treeNeedsReset` in monochromatic search, it is necessary there because we might search with different values for k, which means different bounds.
But, what I mean is that, we can track the reference tree, because it is part of the NeighborSearch class, but we can not track the queryTree.

**Comment by MarcosPividori** [comment]
Maybe it is clearer this way:

"We should reset the stats in the query tree every time we reuse it to do KNN."

Why? Because we could be searching in a different reference dataset, or with a different value of k. Both of these situations mean different values for the bounds.

We have 3 Search() methods in NeighborSearch class:

+ When doing monochromatic search the query tree is properly resetted because 
the query tree is the same than the reference tree. So, everyhing is ok.
```
void Search(const size_t k,
              arma::Mat<size_t>& neighbors,
              arma::mat& distances);
```

+ When we search with different query and reference sets:
   - If we create a query tree inside, we have no problem, because new tree implies proper stats:
```
void Search(const MatType& querySet,
              const size_t k,
              arma::Mat<size_t>& neighbors,
              arma::mat& distances);
```

  - If we receive a query tree, we don't know the state of stats there... This is the special case I am talking about:
```
void Search(Tree* queryTree,
              const size_t k,
              arma::Mat<size_t>& neighbors,
              arma::mat& distances);
```

**Comment by rcurtin** [comment]
I think I have gotten myself confused again.  You are right, there is actually no issue with `treeNeedsReset` since that can only happen for the case of monochromatic search.  So the bug I thought existed, does not exist.

I was not thinking about the overload where the user passes their own query tree; in that case, we have no need to reset the query tree.  I think that we should add documentation ensuring that the user passes a tree with reset bounds, instead of resetting it for them, because (a) that will slow down the `Search()` method and (b) this is already the "advanced" way to call `Search()`, so if users want to build their own special tree that possibly has non-reset bounds for their own reasons, we should allow them to do that.

So in short, I agree with your original statement, that we should add documentation, but I got confused and sidetracked while I was reading through trying to understand. :)
